diff --git a/configure.ac b/configure.ac
index 2d71c838f8..a0cbfa1ee9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2015,7 +2015,7 @@ CURL_WITH_RUSTLS
 
 dnl link required libraries for USE_WIN32_CRYPTO or USE_SCHANNEL
 if test "x$USE_WIN32_CRYPTO" = "x1" -o "x$USE_SCHANNEL" = "x1"; then
-  LIBS="-ladvapi32 -lcrypt32 $LIBS"
+  LIBS="-ladvapi32 -lcrypt32 -lbcrypt $LIBS"
 fi
 
 dnl link bcrypt for BCryptGenRandom() (used when building for Vista or newer)
diff --git a/libcurl.pc.in b/libcurl.pc.in
index 9db6b0f895..327f4d0a0c 100644
--- a/libcurl.pc.in
+++ b/libcurl.pc.in
@@ -39,3 +39,4 @@ Version: @CURLVERSION@
 Libs: -L${libdir} -lcurl @LIBCURL_NO_SHARED@
 Libs.private: @LIBCURL_LIBS@
 Cflags: -I${includedir} @CPPFLAG_CURL_STATICLIB@
+Cflags.private: -DCURL_STATICLIB
