diff --git a/meson.build b/meson.build
index f9fe4e7263..5eaea21968 100644
--- a/meson.build
+++ b/meson.build
@@ -1473,16 +1473,16 @@ endif
 
 
 # macOS features
-macos_sdk_version_py = find_program(join_paths(source_root, 'TOOLS', 'macos-sdk-version.py'),
-                                    required: get_option('swift-build').require(darwin))
-macos_sdk_path = ''
-macos_sdk_version = '0.0'
-if macos_sdk_version_py.found()
+macos_sdk_version_py = find_program(join_paths(source_root, 'TOOLS',
+                                    'macos-sdk-version.py'))
+macos_sdk_info = ['', '0.0']
+if darwin
     macos_sdk_info = run_command(macos_sdk_version_py, check: true).stdout().split(',')
-    macos_sdk_path = macos_sdk_info[0].strip()
-    macos_sdk_version = macos_sdk_info[1]
 endif
 
+macos_sdk_path = macos_sdk_info[0].strip()
+macos_sdk_version = macos_sdk_info[1]
+
 if macos_sdk_path != ''
     message('Detected macOS sdk path: ' + macos_sdk_path)
 endif
