diff --git a/Makerules b/Makerules
index f582dc0614..1902ef73f4 100644
--- a/Makerules
+++ b/Makerules
@@ -379,3 +379,20 @@ ifeq "$(OS)" "ios"
   LD = xcrun ld
   RANLIB = xcrun ranlib
 endif
+
+ifeq "$(OS)" "psvita"
+  CC = arm-vita-eabi-gcc
+  CXX = arm-vita-eabi-g++
+  AR = arm-vita-eabi-ar
+  LD = arm-vita-eabi-ld
+  RANLIB = arm-vita-eabi-ranlib
+  XCFLAGS = $(CPPFLAGS) '-Dtimegm(x)=-1'
+  USE_SYSTEM_FREETYPE = yes
+  USE_SYSTEM_HARFBUZZ = yes
+  SYS_FREETYPE_LIBS =
+  SYS_HARFBUZZ_CFLAGS = $(shell arm-vita-eabi-pkg-config --cflags harfbuzz)
+  SYS_HARFBUZZ_LIBS = $(shell arm-vita-eabi-pkg-config --libs harfbuzz)
+  USE_SYSTEM_LIBJPEG = yes
+  USE_SYSTEM_ZLIB = yes
+  XCFLAGS += -DHAVE_SIGSETJMP=0
+endif
